### DISCUSSION TITLE
[BOOK] 클린 코드 질문

### DISCUSSION THREAD

**Discussion by fkdl0048** [fkdl0048/Mentoring-Project#58] [Q&A] [answered]
## 정리 내용

```java
// 좋은 예
Complex fulcrumPoint = Complex.FromRealNumber(23.0);

Complex fulcrumPoint = new Complex(23.0);
```

*생성자 사용을 제한하려면 해당 생성자를 private로 선언한다.*  

음... 모든 생성자를 private으로 선언하고 정적 팩토리 메서드를 사용하라는 것 같다..?  

이거에 대해서 조금 더 생각해봐야 할 것 같다..  

물론 더 명시적이고 가독성이 좋지만 모든 생성자를 가리고 정적 팩토리 메서드를 사용하는 비용이 조금 생각이 든다..  

선택과 집중을 해야하는 부분일까 아니면 좀 더 좋은 쪽의 결정일까.

## 질문

클래스를 많이 생성하고 세분화 하다 보니 생성자는 별 생각없이 new로 선언해서 사용을 했는데 이러한 정적 팩토리 메서드의 형태로 가져가는 이점은 확실하게 알 것 같지만, 세분화 된 모든 클래스를 저렇게 까지 할 필요가 있을까..? 라는 생각이 듭니다.  

물론 어느정도 분류가 되는 클래스는 정적 팩토림 메서드로 관리하여 생성자를 다양하게 활용 가능하며 네이밍자체도 가독성이 좋은 것 같습니다..

이거에 대해서 계속 고민하게 되어서 종필님의 생각을 듣고 싶습니다..!

**Comment by jongfeel** [comment, marked as answer]
예시로 나온 코드는 정적 메서드로 만들어도 순수 함수로 쓸 수 있기 때문에 그렇습니다.

https://en.wikipedia.org/wiki/Pure_function

객체의 상태를 가지고 있으면서 변화가 일어나지 않는 클래스는 정적 메서드를 통해 순수 함수로 사용하는게 관례로 생각하시면 좋고
생성자를 통해 값(혹은 객체)을 넣은 후 특정 시점에 객체 내부에 변화(객체 메서드 호출, 값 변경)를 통해 그 값이나 변화 시점을 얻기 위해서라면 객체로 관리하는게 좋습니다.

우선은 순수 함수가 저런 형태를 많이 가진다고 이해하시면 좋을 것 같습니다.

**Reply by fkdl0048** [reply]
팩토리같은 내부 변화 없이 반환으로만 처리되는 클래스들은 순수함수로 생성자를 관리..  

이해가 한 80%된 것 같은데 멘토링 때 더 물어보겠습니다..!
